avm/src/daemon/ctrl.rs

```diff
@@ -0,0 +1,200 @@
+  pub fn update(self: &mut LogRendezvousHash, ids: Vec<String>) {
+    let mut sorted_hashes = Vec::with_capacity(ids.len());
+    for id in ids {
+      sorted_hashes.push(HashedId::new(id));
+    }
+    sorted_hashes.sort_by(|a, b| a.hash.cmp(&b.hash));
+    self.sorted_hashes = sorted_hashes;
+  }
+
+  pub fn get_leader_id(&self) -> &str {
+    let mut last_idx = self.sorted_hashes.len() - 1;
+    while last_idx > 0 && !self.sorted_hashes[last_idx].is_up {
+      last_idx = last_idx - 1;
+    }
+    &self.sorted_hashes[last_idx].id
+  }
+
+  pub fn get_mut_nodes(self: &mut LogRendezvousHash) -> &mut Vec<HashedId> {
+    &mut self.sorted_hashes
+  }
+
+  // Runs a binary search for the record whose hash is closest to the key hash without
+  // going over. If none are found, the *last* record in the list is returned as it wraps around.
+  pub fn _get_id_for_key(&self, key: &str) -> &str {
+    let mut key_hasher = XxHash64::with_seed(0xa1a2);
+    key_hasher.write(key.as_bytes());
+    let key_hash = key_hasher.finish();
+    let idx = match self
+      .sorted_hashes
+      .binary_search_by(|a| a.hash.cmp(&key_hash))
+    {
+      Ok(res) => res,
+      // All were too large, implies last (which wraps around) owns it
+      Err(_) => self.sorted_hashes.len() - 1,
+    };
+    &self.sorted_hashes[idx].id
+  }
+}
+
+pub struct ControlPort {
+  lrh: LogRendezvousHash,
+  client: Client<HttpsConnector<HttpConnector>>,
+  // TODO: Once the crazy type info of the server can be figured out, we can attach it to this
+  // struct and then make it possible to wind down the control port server
+  // server: &'a dyn Service<std::convert::Infallible>,
+}
+
+async fn control_port(req: Request<Body>) -> Result<Response<Body>, Infallible> {
+  let cluster_secret = CLUSTER_SECRET.get().unwrap();
+  if cluster_secret.is_some() && !req.headers().contains_key(cluster_secret.as_ref().unwrap()) {
+    // If this control port is guarded by a secret string, make sure there's a header with that
+    // secret as the key (we don't care about the value) and abort otherwise
+    Ok(Response::builder().status(500).body("fail".into()).unwrap())
+  } else if TcpStream::connect("127.0.0.1:443").is_err() {
+    // If the Alan HTTPS server has not yet started, mark as a failure
+    Ok(Response::builder().status(500).body("fail".into()).unwrap())
+  } else if Path::new("./Dockerfile").exists()
+    && Path::new("./app.tar.gz").exists()
+    && TcpStream::connect("127.0.0.1:8088").is_err()
+  {
+    // If this is an Anycloud deployment and the child process hasn't started, mark as a failure
+    // TODO: Any way to generalize this so we don't have special logic for Anycloud?
+    Ok(Response::builder().status(500).body("fail".into()).unwrap())
+  } else {
+    // Everything passed, send an ok
+    Ok(Response::builder().status(200).body("ok".into()).unwrap())
+  }
+}
+
+mod naive {
```

Comment:
we are not putting this in its own file to avoid circular dependencies, right? if so, can we leave a comment about it?

Reply:
Oh, the [original code I based this on](https://github.com/ctz/rustls/blob/51961e1e53380441fb49efdd0e688181a9b275be/rustls-mio/examples/tlsclient.rs#L449-L468) did it that way. I just refactored it until I got it up-to-date and working. I can remove the inner `mod` if you prefer.

Reply:
got it, definitely not something blocking / can happen in the next PR. maybe that is idiomatic rust I am just unfamiliar with it so it looks a bit weird